File: examples/transactions/v1/java/OfflineTransactionAuthorization.java
package com.algorand.javatest;

import java.math.BigInteger;
import java.nio.file.Files;
import java.nio.file.Paths;

import com.algorand.algosdk.account.Account;
import com.algorand.algosdk.algod.client.AlgodClient;
import com.algorand.algosdk.algod.client.api.AlgodApi;
import com.algorand.algosdk.algod.client.auth.ApiKeyAuth;
import com.algorand.algosdk.algod.client.model.TransactionID;
import com.algorand.algosdk.algod.client.model.TransactionParams;
import com.algorand.algosdk.crypto.Address;
import com.algorand.algosdk.crypto.Digest;
import com.algorand.algosdk.transaction.SignedTransaction;
import com.algorand.algosdk.transaction.Transaction;
import com.algorand.algosdk.util.Encoder;

public class OfflineTransactionAuthorization {
    public AlgodApi algodApiInstance = null;

    // utility function to connect to a node
    private AlgodApi connectToNetwork(){

        // Initialize an algod client
        final String ALGOD_API_ADDR = <algod-address>;
        final String ALGOD_API_TOKEN = <algod-token>;

        AlgodClient client = (AlgodClient) new AlgodClient().setBasePath(ALGOD_API_ADDR);
        ApiKeyAuth api_key = (ApiKeyAuth) client.getAuthentication("api_key");
        api_key.setApiKey(ALGOD_API_TOKEN);
        algodApiInstance = new AlgodApi(client);   
        return algodApiInstance;
    }
    // utility function to wait on a transaction to be confirmed    
    public void waitForConfirmation( String txID ) throws Exception{
        if( algodApiInstance == null ) connectToNetwork();
        while(true) {
            try {
                //Check the pending tranactions
                com.algorand.algosdk.algod.client.model.Transaction pendingInfo = algodApiInstance.pendingTransactionInformation(txID);
                if (pendingInfo.getRound() != null && pendingInfo.getRound().longValue() > 0) {
                    //Got the completed Transaction
                    System.out.println("Transaction " + pendingInfo.getTx() + " confirmed in round " + pendingInfo.getRound().longValue());
                    break;
                } 
                algodApiInstance.waitForBlock(BigInteger.valueOf( algodApiInstance.getStatus().getLastRound().longValue() +1 ) );
            } catch (Exception e) {
                throw( e );
            }
        }

    }
    public void writeUnsignedTransaction(){

        // connect to node
        if( algodApiInstance == null ) connectToNetwork();

        final String DEST_ADDR = <transaction-reciever>;
        final String SRC_ADDR = <transaction-sender>;

        try { 
            // Get suggested parameters from the node
            TransactionParams params = algodApiInstance.transactionParams();                     
            BigInteger firstRound = params.getLastRound();
            String genId = params.getGenesisID();
            Digest genesisHash = new Digest(params.getGenesishashb64());

            // create transaction
            BigInteger amount = BigInteger.valueOf(200000);
            BigInteger lastRound = firstRound.add(BigInteger.valueOf(1000));  
            Transaction tx = new Transaction(new Address(SRC_ADDR),  
                    BigInteger.valueOf(1000), firstRound, lastRound, 
                    null, amount, new Address(DEST_ADDR), genId, genesisHash);
            // save as signed even though it has not been
            SignedTransaction stx = new SignedTransaction();
            stx.tx = tx;  
            // Save transaction to a file 
            Files.write(Paths.get("./unsigned.txn"), Encoder.encodeToMsgPack(stx));
            System.out.println("Transaction written to a file");
        } catch (Exception e) { 
            System.out.println("Save Exception: " + e); 
        }

    }
    public void readUnsignedTransaction(){

        try {
            // connect to node
            if( algodApiInstance == null ) connectToNetwork();

            // read transaction from file
            SignedTransaction decodedTransaction = Encoder.decodeFromMsgPack(
                Files.readAllBytes(Paths.get("./unsigned.txn")), SignedTransaction.class);            
            Transaction tx = decodedTransaction.tx;           

            // recover account    
            String SRC_ACCOUNT = <25-word-passphrase>;
            Account src = new Account(SRC_ACCOUNT);

            // sign transaction
            SignedTransaction signedTx = src.signTransaction(tx);
            byte[] encodedTxBytes = Encoder.encodeToMsgPack(signedTx);

            // submit the encoded transaction to the network
            TransactionID id = algodApiInstance.rawTransaction(encodedTxBytes);
            System.out.println("Successfully sent tx with id: " + id);
            waitForConfirmation(id.getTxId());

        } catch (Exception e) {
            System.out.println("Submit Exception: " + e); 
        }


    }
    public void writeSignedTransaction(){

        // connect to node
        if( algodApiInstance == null ) connectToNetwork();

        final String DEST_ADDR = <transaction-reciever>;
        final String SRC_ADDR = <transaction-sender>;;

        try { 

            // Get suggested parameters from the node
            TransactionParams params = algodApiInstance.transactionParams();
            BigInteger firstRound = params.getLastRound();
            String genId = params.getGenesisID();
            Digest genesisHash = new Digest(params.getGenesishashb64());

            // create transaction 
            BigInteger amount = BigInteger.valueOf(200000);
            BigInteger lastRound = firstRound.add(BigInteger.valueOf(1000));  
            Transaction tx = new Transaction(new Address(SRC_ADDR),  
                    BigInteger.valueOf(1000), firstRound, lastRound, 
                    null, amount, new Address(DEST_ADDR), genId, genesisHash);

            // recover account    
            String SRC_ACCOUNT = <25-word-passphrase>;                    
            Account src = new Account(SRC_ACCOUNT);

            // sign transaction
            SignedTransaction signedTx = src.signTransaction(tx);                    

            // save signed transaction to  a file 
            Files.write(Paths.get("./signed.txn"), Encoder.encodeToMsgPack(signedTx));
        } catch (Exception e) { 
            System.out.println("Save Exception: " + e); 
        }

    }

    public void readSignedTransaction(){

        try {
            // connect to a node
            if( algodApiInstance == null ) connectToNetwork();

            //Read the transaction from a file 
            SignedTransaction decodedSignedTransaction = Encoder.decodeFromMsgPack(
                Files.readAllBytes(Paths.get("./signed.txn")), SignedTransaction.class);   
            System.out.println("Signed transaction with txid: " + decodedSignedTransaction.transactionID);           

            // Msgpack encode the signed transaction
            byte[] encodedTxBytes = Encoder.encodeToMsgPack(decodedSignedTransaction);

            //submit the encoded transaction to the network
            TransactionID id = algodApiInstance.rawTransaction(encodedTxBytes);
            System.out.println("Successfully sent tx with id: " + id); 
            waitForConfirmation(id.getTxId());

        } catch (Exception e) {
            System.out.println("Submit Exception: " + e); 
        }


    }
    public static void main(String args[]) throws Exception {
        OfflineTransactionAuthorization mn = new OfflineTransactionAuthorization();
        mn.writeUnsignedTransaction();
        mn.readUnsignedTransaction();

        //mn.writeSignedTransaction();
        //mn.readSignedTransaction();

    }

}

File: examples/transactions/v1/go/offlineTransactionAuthorization.go
package main

import (
    "fmt"
    "io/ioutil"

    "golang.org/x/crypto/ed25519"

    "github.com/algorand/go-algorand-sdk/client/algod"
    "github.com/algorand/go-algorand-sdk/crypto"
    "github.com/algorand/go-algorand-sdk/mnemonic"
    "github.com/algorand/go-algorand-sdk/encoding/msgpack"
    "github.com/algorand/go-algorand-sdk/transaction"
    "github.com/algorand/go-algorand-sdk/types"
)

// Function that waits for a given txId to be confirmed by the network
func waitForConfirmation(algodClient algod.Client, txID string) {
    for {
        pt, err := algodClient.PendingTransactionInformation(txID)
        if err != nil {
            fmt.Printf("waiting for confirmation... (pool error, if any): %s\n", err)
            continue
        }
        if pt.ConfirmedRound > 0 {
            fmt.Printf("Transaction "+pt.TxID+" confirmed in round %d\n", pt.ConfirmedRound)
            break
        }
        nodeStatus, err := algodClient.Status()
        if err != nil {
            fmt.Printf("error getting algod status: %s\n", err)
            return
        }
        algodClient.StatusAfterBlock( nodeStatus.LastRound + 1)
    }
}
// utility function to recover account and return sk and address
func recoverAccount()(string, ed25519.PrivateKey) {
    const passphrase = <your-25-word-mnemonic>

    sk, err := mnemonic.ToPrivateKey(passphrase)
    if err != nil {
        fmt.Printf("error recovering account: %s\n", err)
        return "", nil
    }
    pk := sk.Public()
    var a types.Address
    cpk := pk.(ed25519.PublicKey)
    copy(a[:], cpk[:])
    fmt.Printf("Address: %s\n", a.String()) 
    address := a.String()
    return address, sk 
}
// utility funciton to setup connection to node
func setupConnection()( algod.Client ){
    const algodToken = <algod-token>
    const algodAddress = <algod-address>
    algodClient, err := algod.MakeClient(algodAddress, algodToken)
    if err != nil {
        fmt.Printf("failed to make algod client: %s\n", err)
    }
    return algodClient
}

func saveUnsignedTransaction() {

    // setup connection
    algodClient := setupConnection()

    // recover account for example
    addr, _ := recoverAccount();

    // get network suggested parameters
    txParams, err := algodClient.SuggestedParams()
    if err != nil {
        fmt.Printf("error getting suggested tx params: %s\n", err)
        return
    }

    // create transaction
    toAddr := <transaction-receiver>
    genID := txParams.GenesisID
    tx, err := transaction.MakePaymentTxn(addr, toAddr, 1, 100000,
        txParams.LastRound, txParams.LastRound+100, nil, "", 
        genID, txParams.GenesisHash)
    if err != nil {
        fmt.Printf("Error creating transaction: %s\n", err)
        return
    }
    unsignedTx := types.SignedTxn{
        Txn:  tx,
    }        

    // save unsigned Transaction to file
    err = ioutil.WriteFile("./unsigned.txn", msgpack.Encode(unsignedTx), 0644)
    if err == nil {
        fmt.Printf("Saved unsigned transaction to file\n")
        return
    }
    fmt.Printf("Failed in saving trx to file, error %s\n", err)

}
func readUnsignedTransaction(){

    // setup connection
    algodClient := setupConnection()

    // read unsigned transaction from file
    dat, err := ioutil.ReadFile("./unsigned.txn")
    if err != nil {
        fmt.Printf("Error reading transaction from file: %s\n", err)
        return
    }
    var unsignedTxRaw types.SignedTxn 
    var unsignedTxn types.Transaction

    msgpack.Decode(dat, &unsignedTxRaw)

    unsignedTxn = unsignedTxRaw.Txn

    // recover account and sign transaction
    addr, sk := recoverAccount();
    fmt.Printf("Address is: %s\n", addr)
    txid, stx, err := crypto.SignTransaction(sk, unsignedTxn)
    if err != nil {
        fmt.Printf("Failed to sign transaction: %s\n", err)
        return
    }
    fmt.Printf("Transaction id: %s\n", txid)

    // send transaction to the network
    sendResponse, err := algodClient.SendRawTransaction(stx)
    if err != nil {
        fmt.Printf("failed to send transaction: %s\n", err)
        return
    }
    fmt.Printf("Transaction ID: %s\n", sendResponse.TxID)
    waitForConfirmation(algodClient, sendResponse.TxID)
}


func saveSignedTransaction() {

    // setup connection
    algodClient := setupConnection()

    // recover account
    addr, sk := recoverAccount();

    // get network suggested parameters
    txParams, err := algodClient.SuggestedParams()
    if err != nil {
        fmt.Printf("error getting suggested tx params: %s\n", err)
        return
    }

    // create transaction
    toAddr := <transaction-receiver>
    genID := txParams.GenesisID
    tx, err := transaction.MakePaymentTxn(addr, toAddr, 1, 100000,
        txParams.LastRound, txParams.LastRound+100, nil, "", 
        genID, txParams.GenesisHash)
    if err != nil {
        fmt.Printf("Error creating transaction: %s\n", err)
        return
    }

    // sign the Transaction, msgpack encoding happens in sign
    txid, stx, err := crypto.SignTransaction(sk, tx)
    if err != nil {
        fmt.Printf("Failed to sign transaction: %s\n", err)
        return
    }
    fmt.Printf("Made signed transaction with TxID %s: %x\n", txid, stx)

    //Save the signed transaction to file
    err = ioutil.WriteFile("./signed.stxn", stx, 0644)
    if err == nil {
        fmt.Printf("Saved signed transaction to file\n")
        return
    }
    fmt.Printf("Failed in saving trx to file, error %s\n", err)

}
func readSignedTransaction(){

    // setup connection
    algodClient := setupConnection()

    // read unsigned transaction from file
    dat, err := ioutil.ReadFile("./signed.stxn")
    if err != nil {
        fmt.Printf("Error reading signed transaction from file: %s\n", err)
        return
    }

    // send the transaction to the network
    sendResponse, err := algodClient.SendRawTransaction(dat)
    if err != nil {
        fmt.Printf("failed to send transaction: %s\n", err)
        return
    }

    fmt.Printf("Transaction ID: %s\n", sendResponse.TxID)
    waitForConfirmation(algodClient, sendResponse.TxID)
}
func main() {
    //saveUnsignedTransaction()
    //readUnsignedTransaction()

    saveSignedTransaction()
    readSignedTransaction()

}    